File: backend/app/schemas/content.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class DynamicSectionItemRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    target_type: str = Field(pattern="^(product|restaurant)$")
    product_id: str | None = Field(default=None, max_length=36)
    vendor_id: str | None = Field(default=None, max_length=36)
    sort_order: int = Field(default=0, ge=0, le=9999)


class DynamicSectionCreateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    title: str = Field(min_length=2, max_length=120)
    slug: str = Field(min_length=2, max_length=140, pattern="^[a-z0-9-]+$")
    section_type: str = Field(default="products", pattern="^(products|restaurants|mixed)$")
    is_active: bool = True
    sort_order: int = Field(default=0, ge=0, le=9999)


class DynamicSectionUpdateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    title: str = Field(min_length=2, max_length=120)
    section_type: str = Field(default="products", pattern="^(products|restaurants|mixed)$")
    is_active: bool = True
    sort_order: int = Field(default=0, ge=0, le=9999)


class DynamicSectionItemResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str
    target_type: str
    product_id: str | None
    vendor_id: str | None
    sort_order: int


class DynamicSectionResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str
    title: str
    slug: str
    section_type: str
    is_active: bool
    sort_order: int
    created_at: datetime
    items: list[DynamicSectionItemResponse]


class HomeProductCard(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str
    name: str
    brand: str
    main_image_url: str | None
    is_sponsored: bool
    is_boosted: bool
    amount: float
    currency: str


class HomeRestaurantCard(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str
    name: str
    slug: str


class HomeSectionResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str
    title: str
    slug: str
    section_type: str
    products: list[HomeProductCard]
    restaurants: list[HomeRestaurantCard]


class HomeContentResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    top_banner_url: str | None
    sections: list[HomeSectionResponse]


class AdClickRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    product_id: str = Field(min_length=1, max_length=36)
    section_slug: str | None = Field(default=None, max_length=140)


class AdClickProductStat(BaseModel):
    model_config = ConfigDict(extra="forbid")

    product_id: str
    clicks: int


class AdClickStatsResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    total_clicks: int
    clicks_last_7_days: int
    by_product: list[AdClickProductStat]

File: frontend/src/types/content.ts
export type HomeContentProduct = {
  id: string;
  name: string;
  brand: string;
  main_image_url: string | null;
  amount: number;
  currency: string;
};

export type HomeContentRestaurant = {
  id: string;
  name: string;
  slug: string;
};

export type HomeContentSection = {
  id: string;
  title: string;
  slug: string;
  section_type: "products" | "restaurants" | "mixed";
  products: HomeContentProduct[];
  restaurants: HomeContentRestaurant[];
};

export type HomeContent = {
  top_banner_url: string | null;
  sections: HomeContentSection[];
};

export type SectionItemInput = {
  target_type: "product" | "restaurant";
  product_id?: string;
  vendor_id?: string;
  sort_order: number;
};

export type DynamicSection = {
  id: string;
  title: string;
  slug: string;
  section_type: "products" | "restaurants" | "mixed";
  is_active: boolean;
  sort_order: number;
  created_at: string;
  items: Array<{
    id: string;
    target_type: "product" | "restaurant";
    product_id: string | null;
    vendor_id: string | null;
    sort_order: number;
  }>;
};

export type AdClickStats = {
  total_clicks: number;
  clicks_last_7_days: number;
  by_product: Array<{
    product_id: string;
    clicks: number;
  }>;
};

export type AdminCategory = {
  id: string;
  name: string;
  slug: string;
  parent_id: string | null;
  is_active: boolean;
  created_at: string;
};
